milestones/data.py

```diff
@@ -0,0 +1,390 @@
+# pylint: disable=no-member
+"""
+Application data management/abstraction layer.  Responsible for:
+
+1) Accessing information from various resources:
+* Internal application state  (through local models in models.py)
+* External application state  (ORM bindings with other apps, yuck)
+* Remote data services (through service adapters in resources.py)
+
+2) Calculating derivative information from existing state:
+* Algorithms and data manipulations
+* Aggregations
+* Annotations
+* Alternative data representations
+
+Accepts and returns standard Python data structures (dicts, arrays of dicts)
+for easy consumption and manipulation by callers -- the queryset stops here!
+
+Note the terminology difference at this layer vs. API -- create/fetch/update/delete/
+
+When the time comes for remote resources, import the module like so:
+if getattr(settings, 'TEST_MODE', False) or os.getenv('TRAVIS_MODE', False):
+    import milestones.tests.mocks.resources as remote
+else:
+    import milestones.resources as remote
+"""
+from . import exceptions
+from . import models as internal
+from . import serializers
+
+RELATIONSHIP_TYPES = {
+    'FULFILLS': "fulfills",
+    'REQUIRES': "requires",
+}
+
+
+# PRIVATE/INTERNAL METHODS
+def _get_milestone_relationship_type(relationship):
+    """
+    Retrieves milestone relationship type object from backend
+    """
+    try:
+        return internal.MilestoneRelationshipType.objects.get(
+            name=relationship,
+            active=True
+        )
+    except internal.MilestoneRelationshipType.DoesNotExist:
+        exceptions.raise_exception(
+            "MilestoneRelationshipType",
+            {'name': relationship},
+            exceptions.InvalidMilestoneRelationshipTypeException
+        )
+
+
+# PUBLIC METHODS
+def create_milestone(milestone):
+    """
+    Inserts a new milestone into app/local state
+    Returns a dictionary representation of the object
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    milestone, __ = internal.Milestone.objects.get_or_create(  # pylint: disable=invalid-name
+        namespace=milestone_obj.namespace,
+        name=milestone_obj.name,
+        active=True,
+        defaults={
+            'description': milestone_obj.description,
+        }
+    )
+    return serializers.serialize_milestone(milestone)
```

Comment:
1. Please specify in the docstring what the `milestone` var is.
2. Am I reading this right that if you create a milestone and then soft-delete it (`active=False`), and then go to create it again, it will try to create a new milestone instead of reactivating the old one?